### PR TITLE
Implementing inference api

### PR DESCRIPTION
**Pull request by FelixKirsch**
see:
- https://github.com/code-kern-ai/refinery-weak-supervisor/pull/12
- https://github.com/code-kern-ai/refinery-commercial-proxy/pull/1
- https://github.com/code-kern-ai/refinery-ml-exec-env/pull/15
- https://github.com/code-kern-ai/refinery-tokenizer/pull/13
- https://github.com/code-kern-ai/refinery-gateway/pull/84
- https://github.com/code-kern-ai/refinery-embedder/pull/21
- https://github.com/code-kern-ai/refinery-inference-exec-env

### CODE REVIEW

File: controller.py [hunk 1]
```diff
@@ -485,6 +496,12 @@ def delete_embedding(project_id: str, embedding_id: str) -> int:
     org_id = organization.get_id_by_project_id(project_id)
     s3.delete_object(org_id, project_id + "/" + object_name)
     request_util.delete_embedding_from_neural_search(embedding_id)
+    if get_config_value("is_managed"):
```
Comment (JWittmeyer) [inline controller.py]: absolute edge case for dev but i don't think we need the manged check in the delete methods since if the path exists it should be deleted anyway. (For the dev who switches between managed and not :) )

File: start [hunk 2]
```diff
@@ -27,6 +27,20 @@ then
...
 
+INFERENCE_DIR=${PWD%/*}/dev-setup/inference/
+if [ ! -d "$_DIR" ] 
+then
+    INFERENCE_DIR=${PWD%/*/*}/dev-setup/inference/
+    if [ ! -d "$INFERENCE_DIR" ] 
+    then
+        # to include volume for local development, use the dev-setup model data folder:  
+        # alternative use manual logic with
+        # -v /path/to/dev-setup/model-data:/models \
```
Comment (JWittmeyer) [inline start]: copy coment

File: controller.py [hunk 3]
```diff
@@ -449,6 +450,16 @@ def run_encoding(
...
             )
+
+        if get_config_value("is_managed"):
+            pickle_path = os.path.join(
+                "/inference", request.project_id, f"embedder-{embedding_id}.pkl"
+            )
+            if not os.path.exists(pickle_path):
+                os.makedirs(os.path.dirname(pickle_path), exist_ok=True)
+                with open(pickle_path, "wb") as f:
+                    pickle.dump(embedder, f)
```
Comment (JWittmeyer) [inline controller.py:461]: would it be possible to have an endpoint to store this? Not sure what happens with the embedded but recreating the whole embedding just to store it feels like a lot of time for 50k projects.

**Review by JWittmeyer — APPROVED** [review]

File: controller.py (hunk 1 above)
Comment (FelixKirsch) [reply controller.py]: removed the if-clause

File: start (hunk 2 above)
Comment (FelixKirsch) [reply start]: changed

File: controller.py (hunk 3 above)
Comment (FelixKirsch) [reply controller.py:461]: The critical part of storing the embedder is the included data for the Principal Component Analysis, which is used to reduce the dimensionality of the created embeddings. The PCA is calculated on a subset of the embeddings (I think the first 5 batches with 128 embeddings each). But I don't think we would know which embeddings have been used for the PCA. However, I believe the deviation by choosing the embeddings for the PCA random would be neglectable.

We would only need to save the PCA for each project (embedding per project). The actual embedders could be shared between projects. This would also save disk space. It is also part of the optimization list.

So, in short, yes this should be possible, but would require some changes.

Comment (JWittmeyer) [reply controller.py:461]: works for me, :)
